XLS-85d Token Escrow

dangell7 opened the pull request.

sappenin commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+**Failure Conditions:**
+
+1. **Source Not Authorized to Hold Token:**
+   - If authorization is required and the source is not authorized, transaction fails with `tecNO_AUTH`.
```
Is there anything we can do to mitigate this condition? This feels like a stuck escrow (which may be unavoidable, but would be nice to avoid). In this case, the sender may not be able to rectify the situation.

dangell7 replied:
Not sure, I think if the holder is not authorized and there is require auth then allowing them to hold the token violates the require auth.

sappenin replied:
>  if the holder is not authorized and there is require auth then allowing them to hold the token violates the require auth.

Looking into this for MPTs at least, if an account has a non-zero balance of MPT, [then it can't be unauthorized](https://xrpl.org/docs/references/protocol/transactions/types/mptokenauthorize#mptokenauthorize-flags). The issuer would have to Clawback all MPT first, and then un-authorize.

I think we should strive for the same behavior here (i.e., if an issuer is trying to unauthorize an account that has MPT in escrow, then this transaction should fail). Primary reason is it's very inelegant to have functionality work differently across features.

 To pull this off though (at least for MPT), we'd need to perhaps track the total amount of MPT in-escrow on the [MPToken object](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-multi-purpose-tokens#2122-fields) of the sender (or a boolean maybe like `hasEscrow`?). Alternatively, the transaction check would have to iterate through an account's escrows, checking for any MPT that matches the issuer doing the un-auth.

dangell7 replied:
You could store the escrowed balance on the MPT. So the MPTAmount is 0 but the sfEscrowAmount != 0. 

Done here: https://github.com/XRPLF/rippled/pull/5185/commits/3e029b67037eba95058469c30aa0560239e516dd

sappenin replied:
@dangell7 Does the spec needs to be updated to conform with the final implementation? If not, this one can be resolved.

sappenin commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+|-------------------------------|--------------------------------------------------------------------------|----------------------------------------------------------------------------|
+| **Trustlines**                | Required between accounts and issuer                                     | Not used                                                                   |
+| **Issuer Flag for Escrow**    | `lsfAllowTokenEscrow` (account flag)                                       | `tfMPTCanEscrow` (token flag)                                              |
+| **Transfer Flags**            | N/A                                                                      | `tfMPTCanTransfer` must be enabled for transfer and escrow                 |
```
```suggestion
| **Transfer Flags**            | N/A                                                                      | `tfMPTCanTransfer` must be enabled for  escrow                 |
```

theotherian commented:
Any chance this would be added to the `gateway_balances` resource to avoid having to make two trips to ledger to find out the total supply at an issuer?

theotherian commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+**State Changes:**
+
+- **Transfer from Source to Issuer Holding:**
+  - **IOU Tokens**: Amount deducted from the source's trustline balance.
```
This would definitely happen when traversing trustlines via the `account_lines` resource. Seems you'd have to both traverse those and look up the account objects.

I mentioned this in another comment already, but this would make sense to add to the `gateway_balances` resource since that's a one stop shop for the details around supply of a token issued via trustlines.

dangell7 replied:
I think this is an easy fix for `gateway_balances` and `account_lines`. We can added `escrowed` field to gateway and then for account_lines we could add a filter using the same field `escrowed`: true | false?

We should, could also do something with MPT?

dangell7 replied:
This is the result of the "gateway_balances". Currently both the MPT and IOU solutions reduce the outstanding/obligations. For MPT its a big issue because there is a limit, on IOU there is no limit, but I'm wondering if both solutions should not reduce the outstanding/obligations amount.

```
{
	"result" : 
	{
		"account" : "r9QxhA9RghPZBbUchA9HkrmLKaWvkLXU29",
		"escrowed" : 
		{
			"USD" : "1000"
		},
		"ledger_current_index" : 9,
		"obligations" : 
		{
			"USD" : "9000"
		},
		"status" : "success",
		"validated" : false
	}
}
```

For MPT I was thinking something like: 

```
"result" : 
	{
		"index" : "DE87AC023DF34314150963F2EFC5787748B2366F6AAB5DE061E0C9290A6C5259",
		"ledger_current_index" : 10,
		"node" : 
		{
			"Flags" : 40,
			"Issuer" : "rExKpRKXNz25UAjbckCRtQsJFcSfjL9Er3",
			"LedgerEntryType" : "MPTokenIssuance",
			"OutstandingAmount" : "9000",
			"EscrowedAmount" : "1000",
			"OwnerNode" : "0",
			"PreviousTxnID" : "B6486FBC988B2D5E22DD3A40F1F83355032800C1A0515B8E46A8E654D9E3FA88",
			"PreviousTxnLgrSeq" : 9,
			"Sequence" : 4,
			"index" : "DE87AC023DF34314150963F2EFC5787748B2366F6AAB5DE061E0C9290A6C5259",
			"mpt_issuance_id" : "00000004A407AF5856CCF3C42619DAA925813FC955C72983"
		},
		"status" : "success",
		"validated" : false
	}
````

dangell7 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+    - A new trustline or MPT holding cannot be created during `EscrowFinish` if authorization is required.
+- **Freeze and Lock Conditions:**
+  - **IOU Tokens**:
+    - **Deep Freeze**: If the token is deep frozen, the transaction fails with `tecFROZEN`.
```
Only deep freeze but whatever is needed for compliance. Happy to remove.

dangell7 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+
+   - **MPTs**:
+     - **Lock Conditions (Deep Freeze Equivalent)**: The transaction succeeds, allowing the escrow to be cancelled.
+     - **Transfer Flags**: If `tfMPTCanTransfer` is not enabled, transaction fails with `tecNO_PERMISSION`.
```
This is interesting: (except possibly when an escrow is created with source + destination as the same address). 

Yes I agree, you should not need it. I will remove it.

mvadari replied:
I'm curious what you guys think should happen here if this is an escrow to yourself.

sappenin replied:
It's a good point @mvadari -- though at the same I can't really come up with a good reason to disallow this. 

Plus, current XRP escrows allow escrowing to oneself, so making tokens do something different feels like violating the principle of least surprise. 

Also, perhaps being able to demonstrate some sort of "lock up" of one's own funds is one use-case? Ripple escrows come to mind, so in this context maybe putting funds on hold for a time has some value for a business contract or otherwise. Or, maybe something more simple just like, "I decided to put some MPT into an escrow so I don't spend it for 6 months because I'm not great at spending control"?

sappenin replied:
@dangell7 Does the spec needs to be updated to conform with the final implementation? If not, this one can be resolved.

dangell7 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+2. **Source Account Not Authorized to Hold Token:**
+   - If the issuer requires authorization and the source is not authorized, the transaction fails with `tecNO_AUTH`.
+
+3. **Source Account's Token Holding Issues:**
```
Yeah I think `tecUNFUNDED` is fine to use as well

sappenin replied:
Resolving this per my other comments. No changes needed.

dangell7 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+
+- **Transfer from Source to Issuer Holding:**
+  - **IOU Tokens**: Amount deducted from the source's trustline balance.
+  - **MPTs**: Amount moved to a holding controlled by the issuer; the issuer's total outstanding amount remains unchanged.
```
I will have to review the `rippleCredit` function again. I thought the funds go back to the issuer.

dangell7 replied:
The wording could be changed though. "A holding controlled by the issuer" isn't great haha

sappenin replied:
> I will have to review the rippleCredit function again. I thought the funds go back to the issuer.

Ah, I see -- I was imagining a different design where `rippleCredit` is not used (I've looked more closely at the [current code PR](https://github.com/XRPLF/rippled/pull/5185/files) so I'm tracking a bit better now).

So, I see that the current implementation is taking MPT from the sender, and sending it back to the MPT issuer via [here](https://github.com/Transia-RnD/rippled/blob/iou-escrow/src/xrpld/app/tx/detail/Escrow.cpp#L550) (i.e., the issuer is specified as the destination). However, I'm uncertain about this design for two reasons. First, MPT issuers aren't allow to hold MPTs like regular accounts can hold MPT (MPT issuers [don't have `MPToken` objects](https://github.com/XRPLF/XRPL-Standards/tree/master/XLS-0033d-multi-purpose-tokens#a18-why-is-there-no-mptokenredeem-transaction)). This means that when a normal MPT payment is made to an issuer, it's actually just a `redeem` operation, and `outstandingAmount` is the thing that gets reduced. However, in the case of an `EscrowCreate`, we don't want to mutate the `outstandingAmount`. Second, it's unclear what value trying to have an MPT Issuer hold a temporary `MPToken` when we can simply store the Escrowed MPT amount directly on the Escrow object.

So, for MPT escrow, I wonder if we just do something similar to [what happens for XRP](https://github.com/XRPLF/rippled/pull/5185/files#diff-35a70dda86b5499425f5a8ed0b8a236c0623bb785a2f610c18943b29987e163dR541) (i.e., reduce the sender, and increase the amount on the escrow).

This raises another interesting question: Is it acceptable for an MPT issuer to create an escrow? I think this should be fine, but I need to think a bit more about that one.

Thoughts?

sappenin replied:
Agreed that's how `rippleCredit` works. But the Escrow::doApply in your implementation [here](https://github.com/Transia-RnD/rippled/blob/iou-escrow/src/xrpld/app/tx/detail/Escrow.cpp#L550) is sending escrowed funds back to the issuer (that's item #3 in your list, unless I'm misreading your implementation?)

dangell7 replied:
No you're correct. I was stuck on "sending escrowed funds" vs an adjustment to outstanding.

I have updated the code. Added a function called `rippleEscrow`. This now updates a new field `sfEscrowedAmount` and doesn't adjust `sfOutstandingAmount`. 

I was going to add a bool to `rippleCredit` but I think a seperate function is better so that I dont have to guard existing functionality with a rule and it can be used later and maybe its more intuitive. 

https://github.com/XRPLF/rippled/pull/5185/commits/3e029b67037eba95058469c30aa0560239e516dd

sappenin replied:
@dangell7 Does the spec needs to be updated to conform with the final implementation? If not, this one can be resolved.

dangell7 commented:
> Any chance this would be added to the `gateway_balances` resource to avoid having to make two trips to the ledger to find out the total supply at an issuer?

https://github.com/XRPLF/XRPL-Standards/pull/272#discussion_r1984565793

nkramer44 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+    - `CancelAfter`: When the Escrow Expires (Required on IOU/MPT)
+    - `Amount`: Tokens held in escrow.
+    - `TransferRate`: `TransferRate` (IOUs) or `TransferFee` (MPTs) at creation.
+    - `IssuerNode`: Reference to the issuer’s ledger node if applicable.
```
Is `IssuerNode` a pointer to the issuer's `AccountRoot` entry, or does it point to something else?

dangell7 replied:
It points to the page where the escrow keylet was stored in the issuer directory

nkramer44 replied:
Ah ok, that was actually my next question -- it sounds like all Escrows involving an issuer have their keylets stored in the issuer's owner directory, is that correct? Going back to [this thread](https://github.com/XRPLF/XRPL-Standards/pull/272#discussion_r1979979181), I think this means you could compute the total amount of IOUs issued by an issuer by iterating over the issuer's owner directory, summing up all RippleState balances and Escrowed amounts. Is my understanding correct there?

dangell7 replied:
Yes, if the issuer is the sender or the receiver there is a keylet stored in the issuers directory. 

As for the issuer's `escrowed` amount I'm not certain of the exact implementation until I write it...

sappenin replied:
@nkramer44 Can this be resolved?

sappenin replied:
Bump @nkramer44

sappenin replied:
Resolving for now; we can revisit this later if desired.

dangell7 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+     - If the `MPTokenIssuance` of the token being escrowed lacks the `lsfMPTCanTransfer` flag, the transaction fails with `tecNO_PERMISSION`.
+
+2. **Source Account Not Authorized to Hold Token:**
+   - If the issuer requires authorization and the source is not authorized, the transaction fails with `tecNO_AUTH`.
```
Not sure I agree with this, everywhere else this is a `tecNO_AUTH`. Even in the MPT Dex implementation, unless it was changed this result would return `tecNO_AUTH`.

sappenin replied:
Hmmm... I would agree with you for EscrowFinish (if an account is not authorized, then trying to finish an Escrow to that account should result in `tecNO_AUTH`).

However, in this case, an Escrow is being created by a user who has "no tokens". So, the first check shouldn't be, "is a user authorized to hold a token", but instead the first check should simply be, "does a user even have the token?"

With that reasoning, I propose the following:

1. For EscrowCreate --> `tecUnfunded`
2. For EscrowFinish --> `tecNO_AUTH`

(That said, perhaps you could provide a link to the MPT DEX implementation? It's sound off to me that if one were to try to create an offer without holding a token, the first check would be an auth check. Seems like it should first check existence).

sappenin replied:
Nevermind, I've changed my mind. Resolving this per my other comment.

dangell7 commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+
+3. **Source Account's Token Holding Issues:**
+   - **IOU Tokens**: If the source lacks a trustline with the issuer, the transaction fails with `tecNO_LINE`.
+   - **MPTs**: If the source does not hold the MPT, the transaction fails with `tecOBJECT_NOT_FOUND`.
```
Same here, this is taken directly from the MPT Dex implementation and other implementations of MPT. I think this should remain.

sappenin replied:
IDK -- can you imagine trying to use this feature and getting `tecOBJECT_NOT_FOUND`? Anyone unfamiliar with the inner workings of MPT (i.e., most developers) would just be confused. It's seems so much more useful to return `tecUNFUNDED`.

sappenin replied:
Alright, looking at the implementation of CreateOffer and Payment for MPTs a bit more, I think there's too much prior art to change this pattern. I'll resolve these proposed changes and we can keep things as-is.

sappenin commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,243 @@
+     - If the token lacks the `tfMPTCanTransfer` flag, the transaction fails with `tecNO_PERMISSION`.
+
+2. **Source Account Not Authorized to Hold Token:**
+   - If the issuer requires authorization and the source is not authorized, the transaction fails with `tecNO_AUTH`.
```
Resolving per prior-art with MPT Payments and DEX implementations.

dangell7 commented:
Update the spec.

- Cleaned up some redundancy
- Added extra information for StateChanges (OutstandingAmount & EscrowedAmount)
- Added failure for _Issuer as Source_
- Added Future Considerations
- Added Update to Ledger Entry Objects

Open Issues:

Nomenclature for `sfEscrowedAmount`

mvadari commented on XLS-0085d-token-escrow/README.md:
```diff
@@ -0,0 +1,231 @@
+Contributors:
+    Richard Holland, XRPL-Labs [RichardAH](https://github.com/RichardAH)
+    Ed Hennis, Ripple [ximinez](https://github.com/ximinez)
+    Mayukha Vadari, Ripple [msvadari](https://github.com/msvadari)
```
...did I contribute to this spec?

dangell7 replied:
I went ahead and just removed the "contributors"

dangell7 commented:
@dangell7 Add the transfer fee/rate functionality to the EscrowFinish and EscrowCancel sections

ckeshava commented:
1. Why is an `Issuer` forbidden from creating an `Escrow` with their token? I could not foresee problems that allows this design.

In the current world:
- Central Banks lend their currency to other banks (or) large infrastructure projects (or) public-spending welfare programs.
- Corporations reward employees with RSUs. These units have a time-based constraint before they can be `EscrowFinish`'ed

Both these examples have an `Issuer` as the creator of the `Escrow`. 

Furthermore, the use of a `tem...` code (for ex: `temNO_PERMISSION`) would have failed the transaction faster. It would have also saved the fees consumed by a `tecNO_PERMISSION` error code. Why did you choose the latter?

2. The cpp implementation uses `sfLockedAmount` in two MPT-related ledger-entries. Is that equivalent to the `sfEscrowAmount` as mentioned in this specification?

It would be very helpful if the documentation could explicitly state that `sfLockedAmount` is a base-10 specified value (although it uses UINT64 rippled type) and not a base-16 value. (I'm gleaning this based on the usage of `SField::sMD_BaseTen` in its declaration)

dangell7 commented:
> 1. Why is an `Issuer` forbidden from creating an `Escrow` with their token? I could not foresee problems that allows this design.
> 
> In the current world:
> 
> * Central Banks lend their currency to other banks (or) large infrastructure projects (or) public-spending welfare programs.
> * Corporations reward employees with RSUs. These units have a time-based constraint before they can be `EscrowFinish`'ed
> 
> Both these examples have an `Issuer` as the creator of the `Escrow`.
> 
> Furthermore, the use of a `tem...` code (for ex: `temNO_PERMISSION`) would have failed the transaction faster. It would have also saved the fees consumed by a `tecNO_PERMISSION` error code. Why did you choose the latter?
> 
> 2. The cpp implementation uses `sfLockedAmount` in two MPT-related ledger-entries. Is that equivalent to the `sfEscrowAmount` as mentioned in this specification?
> 
> It would be very helpful if the documentation could explicitly state that `sfLockedAmount` is a base-10 specified value (although it uses UINT64 rippled type) and not a base-16 value. (I'm gleaning this based on the usage of `SField::sMD_BaseTen` in its declaration)

I have updated the spec for the sfLockedAmount and `CancelAfter`

https://github.com/XRPLF/XRPL-Standards/pull/292

As for why Issuer cannot create Escrow that was a design decision that was made. I agree it could be valuable for the Issuer to create an escrow but the issuer could also just create a new account, send value and escrow from there. 

As for the tecNO_PERMISSION error response you are right, it should have been a tem error as its more of a transaction malformed error than a ledger error.

dangell7 commented:
> Do you have thoughts on dealing with "orphaned" escrow ledger-objects? Here's a thought experiment:
> 
> * `A` (Escrow creator) creates an escrow with `USD` `IOU` token to destination `B` (Escrow Destination). It has the necessary authorization to hold the IOU.
> * After certain time, the Issuer of the `USD` token decides to remove the trustline-authorization for both `A` and `B`.
> * Neither `EscrowFinish` nor `EscrowCancel` is feasible at this juncture, because `USD` tokens cannot be transferred to either parties.
> * The EscrowCreator `A` is unable to get back its reserve that was taken at the time of escrow-creation.
> 
> Is this scenario discussed elsewhere in the comments, I missed it.

Are you suggesting we add the authorized trustline automatically for the account?

I’m happy to make any changes you would like to recommend.

dangell7 commented:
> > So first, the issuer cannot unauthorize an IOU.
> 
> That is correct, I was considering the case where the Issuer deep-freezes the escrow-creator and/or escrow-destination. A deep-freeze of the trust line will freeze the trust-line balance.
> 
> > maybe just returning the tokens back to the account on Cancel and then rejecting on Finish just like DeepFreeze is the best option
> 
> Are you suggesting returning the tokens back to the issuer in the above line? that seems reasonable, yes.
> 
> I endorse your idea of `TokenEscrow` as "DeletionBlockers", with respect to the Deletion of Trustlines. I also want to draw your attention to [an existing issue](https://github.com/XRPLF/rippled/issues/4698) pertaining to the "default state" of authorized trust lines. I'm hoping you won't be flummoxed by it.

Technically the issuer is never the account so no, it would go back to the user who held the tokens in the first place. But without authorization they wouldn't be able to do anything with them. However, with the new changes, you can't delete a trustline while having an escrow so you would never be able to remove the auth in the first place.

Removing the ability to remove a trustline if you have an escrow solves most of the issues, same with the MPT side. If you have any tokens escrowed you cannot un auth the MPT. 

I think that solves most of the orphan issues.

mvadari commented:
(In the future let's try to have these conversations _before_ the amendment is released, so that we don't need a separate fix amendment)
